### PR TITLE
don't use no_sync when deepspeed doesn't support it for certain zero stages

### PR DESCRIPTION
**Pull request by winglian**
# What does this PR do?
Deepspeed 0.16 has assertions preventing the use of no_sync with zero 2/3. see https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/engine.py#L1986-L2004

it seems people are reporting this here https://github.com/microsoft/DeepSpeed/issues/6793, and I'm assuming that everyone is using accelerate/transformers as downgrading to deepspeed 0.15.4 makes it "work" for them.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@muellerzr 

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by winglian** [comment]
might have to broaden deepspeed for all zero cases? https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/engine.py#L2208-L2209

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -2513,9 +2513,13 @@ def _inner_training_loop(
...
 
                     # We explicitly want to avoid relying on `accelerator.accumulate` for generation training
+                    disable_deepspeed_no_sync = (
+                        self.accelerator.distributed_type == DistributedType.DEEPSPEED
+                        and self.accelerator.deepspeed_engine_wrapped.engine.zero_optimization_partition_gradients()
+                    )
                     context = (
                         functools.partial(self.accelerator.no_sync, model=model)
-                        if i != len(batch_samples) - 1
+                        if i != len(batch_samples) - 1 and not disable_deepspeed_no_sync
```
Comment (muellerzr) [inline src/transformers/trainer.py:2522]:
```suggestion
                        if i != len(batch_samples) - 1 and not self.accelerator.distributed_type == DistributedType.DEEPSPEED
```

**Review by muellerzr — COMMENTED** [review]
Since we should disable it for *all* DS (apparently), let's just go ahead and do that. I'll apply a similar fix in Accelerator. 

cc @SunMarc

**Review by SunMarc — APPROVED** [review]
LGTM with zach suggestion !

**Review by muellerzr — APPROVED** [review]
Thanks! We've also confirmed this fixes up all the fails users reported wrt deepspeed. cc @ArthurZucker for final post wing doing quality ;)

**Review by ArthurZucker — APPROVED** [review]
sorry for breaking this ... and thanks for the fix!

File: src/transformers/trainer.py (hunk 1 above)
Comment (chuangzhidan) [reply src/transformers/trainer.py]: may i ask which version of transformers support this fix-up? mine is 4.46.0. same problem with deepspeed 0.16

**Comment by AetherPrior** [comment]
Strangely, this issue still exists on `deepspeed==0.16.2`, has this fix been pushed to a stable release yet?

**Comment by ArthurZucker** [comment]
This is on 4.48 !

**Comment by fangpings** [comment]
With transformers==4.48.0, accelerate==1.2.1 and deepspeed==0.16.3, still see this issue

**Comment by jianguoz** [comment]
Hi @ArthurZucker @muellerzr , we still face the same issue with transformers==4.48.0/4.49.0, accelerate==1.2.1 and deepspeed==0.16.3. Could you check this?

```text
error: no_sync context manager is incompatible with gradient partitioning logic of ZeRO stage 3
```

**Comment by jianguoz** [comment]
Hi @SunMarc , Below is the my zero 3 config and  error output. It can only work with `gradient_accumulation=1`. I believe people will face same issue when training a model using the latest transformer and deepspeed. 

```
[2025-03-04 19:03:35,551] [INFO] [config.py:991:print_user_config]   json = {
    "bf16": {
        "enabled": true
    }, 
    "optimizer": {
        "type": "AdamW", 
        "params": {
            "lr": 2e-05, 
            "weight_decay": 0
        }
    }, 
    "scheduler": {
        "type": "WarmupDecayLR", 
        "params": {
            "warmup_min_lr": 1e-06, 
            "warmup_max_lr": 2e-05, 
            "warmup_num_steps": 100, 
            "total_num_steps": 1.259000e+03
        }
    }, 
    "zero_optimization": {
        "stage": 3, 
        "overlap_comm": true, 
        "contiguous_gradients": true, 
        "sub_group_size": 1.000000e+09, 
        "reduce_bucket_size": 2.621440e+07, 
        "stage3_prefetch_bucket_size": 0, 
        "stage3_param_persistence_threshold": 5.120000e+04, 
        "stage3_max_live_parameters": 1.000000e+09, 
        "stage3_max_reuse_distance": 1.000000e+09, 
        "stage3_gather_16bit_weights_on_model_save": true
    }, 
    "gradient_accumulation_steps": 3, 
    "gradient_clipping": 1.0, 
    "steps_per_print": inf, 
    "train_batch_size": 48, 
    "train_micro_batch_size_per_gpu": 2, 
    "wall_clock_breakdown": false, 
    "fp16": {
        "enabled": false
    }
}
```

output 
```
step = 0 -- rank 0: error: no_sync context manager is incompatible with gradient partitioning logic of ZeRO stage 3
  0%|                                                                                                                                     | 1/10072 [01:15<210:10:00, 75.13s/it]step = 1 -- rank 0: error: step = 0 -- rank 1: error: no_sync context manager is incompatible with gradient partitioning logic of ZeRO stage 3
no_sync context manager is incompatible with gradient partitioning logic of ZeRO stage 3
  0%|                                                                                                                                     | 1/10072 [01:15<210:30:16, 75.25s/it]step = 0 -- rank 2: error:step = 0 -- rank 5: error:  step = 0 -- rank 3: error:no_sync context manager is incompatible with gradient partitioning logic of ZeRO stage 3no_sync context manager is incompatible with gradient partitioning logic of ZeRO stage 3step = 0 -- rank 6: error: 
step = 0 -- rank 7: error:
no_sync context manager is incompatible with gradient partitioning logic of ZeRO stage 3  
no_sync context manager is incompatible with gradient partitioning logic of ZeRO stage 3no_sync context manager is incompatible with gradient partitioning logic of ZeRO stage 3

step = 0 -- rank 4: error: no_sync context manager is incompatible with gradient partitioning logic of ZeRO stage 3
step = 1 -- rank 1: error: no_sync context manager is incompatible with gradient partitioning logic of ZeRO stage 3
  0%|                                                                                                                                     | 1/10072 [01:15<210:23:32, 75.21s/it]step = 1 -- rank 2: error: step = 1 -- rank 5: error:no_sync context manager is incompatible with gradient partitioning logic of ZeRO stage 3 
no_sync context manager is incompatible with gradient partitioning logic of ZeRO stage 3
step = 1 -- rank 4: error: no_sync context manager is incompatible with gradient partitioning logic of ZeRO stage 3
step = 1 -- rank 7: error:step = 1 -- rank 6: error:  no_sync context manager is incompatible with gradient partitioning logic of ZeRO stage 3no_sync context manager is incompatible with gradient partitioning logic of ZeRO stage 3
```

**Comment by jianguoz** [comment]
Hi @SunMarc , do we have updates for this issue? I saw others have same issues on #34984

**Comment by SunMarc** [comment]
cc @XuehaiPan

**Comment by jianguoz** [comment]
Hi @SunMarc , any update for this incompatible issues between Huggingface and Deepspeed?

**Pull request merged**